client/components/ArtworkForm.js

```diff
@@ -0,0 +1,149 @@
+import React, { Component } from 'react'
+import { Redirect } from 'react-router-dom'
+import { connect } from 'react-redux'
+import { addArtwork, editArtwork, getArtwork } from '../redux/artwork'
+import { mediums } from '../../server/constants'
+
+class ArtworkForm extends Component {
+  constructor(props) {
+    super(props)
+    this.state = {
+      redirect: false,
+      title: '',
+      description: '',
+      year: '',
+      price: '',
+      quantity: '',
+      //genres: [],
+      medium: '',
+      artistId: ''
+    }
+    this.handleChange = this.handleChange.bind(this)
+    //this.handleGenreChange = this.handleGenreChange.bind(this)
+    this.handleSubmit = this.handleSubmit.bind(this)
+  }
+
+  componentDidMount() {
+    if (!this.props.artwork && this.props.match.params.id) {
+      this.props.getArtwork(this.props.match.params.id)
+    }
+    if (this.props.isEditing) {
+      const { title, description, year, price, quantity, medium, artistId } = this.props.artwork
+      this.setState({title, description, year, price, quantity, medium, artistId})
+    }
+  }
+
+  handleChange(ev) {
+    const { target: { name, value } } = ev
+    this.setState({ [name]: value })
+  }
+
+  /*handleGenreChange(ev) {
+    const { genres } = this.state
+    console.log(genres, ev.target)
+  }*/
+
+  handleSubmit(ev) {
+    ev.preventDefault()
+    try {
+      if (this.props.isEditing) {
+        this.props.editArtwork(this.props.artwork.id, this.state)
+      } else {
+        const form = {}
+        Object.entries(this.state).forEach(([key, value]) => {
+          if (value) form[key] = value
+        })
+        this.props.addArtwork(form)
+      }
+      this.setState({redirect: true})
+    } catch (err) {
+      console.log(err)
+    }
+  }
+
+  /*
+    To-do: handle adding genres thru associations.
+  */
+  render() {
+    const { title, description, year, price, quantity, redirect, medium, artistId  } = this.state
+    const { isEditing, artwork, artists, genres } = this.props
+    return redirect ?
+      <Redirect to={isEditing ? `/artworks/${artwork.id}` : '/'} />
+      : (
+      <div className="form-container">
+        <h2>{isEditing ? `Edit ${artwork.title}` : 'Add New Art to Collection'}</h2>
+          {!!isEditing && <img src={this.props.artwork.shopImages.length
+            ? this.props.artwork.shopImages[0].imageURL
+            : '/img/default.jpg'} />
+          }
+
+        <form id="artwork-form" onSubmit={this.handleSubmit}>
+          <label htmlFor="title">Title</label>
+          <input name="title" type="text" value={title} required onChange={this.handleChange} />
+
+          <label htmlFor="artistId">Artist</label>
+            <select name="artistId" value={artistId} onChange={this.handleChange}>
+            <option value="">Other</option>
```

Comment:
If you go straight to the admin page, you never get the artists, which means you can't select one from the drop down. I ran into similar problems doing stuff on the cart.
I worked around it within the component, and you could here, too, but maybe we should have 'a bigger think' about how and when we get data, so that every component can always be guaranteed to have access to certain stuff.

Reply:
yes, good point!